### PR TITLE
Remove unused set-discriminant statements and assignments regardless of rvalue

### PR DESCRIPTION
**Pull request by tmiasko**
* Represent use counts with u32
* Unify use count visitors
* Change RemoveStatements visitor into a function
* Remove unused set-discriminant statements
* Use exhaustive match to clarify what is being optimized
* Remove unused assignments regardless of rvalue kind

### CODE REVIEW

File: compiler/rustc_mir/src/transform/simplify.rs [hunk 1]
```diff
@@ -363,14 +356,14 @@ impl<'tcx> MirPass<'tcx> for SimplifyLocals {
...
+    used_locals: &UsedLocals,
 ) -> IndexVec<Local, Option<Local>> {
     let mut map: IndexVec<Local, Option<Local>> = IndexVec::from_elem(None, &*local_decls);
     let mut used = Local::new(0);
-    for (alive_index, count) in used_locals.iter_enumerated() {
-        // The `RETURN_PLACE` and arguments are always live.
-        if alive_index.as_usize() > arg_count && *count == 0 {
+
+    for alive_index in local_decls.indices() {
+        // The `RETURN_PLACE` and arguments are always used.
```
[the remark this answers is not in the document]
Comment (tmiasko) [reply compiler/rustc_mir/src/transform/simplify.rs]: Slightly reworded the comment.

**Review by wesleywiser — APPROVED** [review]
This looks great and I'm glad to see tests covering the new behavior!

**Comment by tmiasko** [comment]
Added overflow-checks=off to the test and updated the output.

**Comment by Mark-Simulacrum** [comment]
Performance results, as expected, are an [improvement](https://perf.rust-lang.org/compare.html?start=fd542592f08ca0d1f7255600115c2eafdf6b5da7&end=a4d30a7b490065f0aa56f58e508a11546445aea9&stat=instructions:u) on the ctfe stress test. That said, there does appear to be a regression on the cranelift codegen benchmark in LLVM -- perhaps these weren't quite unused? Any thoughts here?

**Comment by tmiasko** [comment]
Those changes are completely boring from LLVM perspective. They remove dead code and affect size estimates accordingly. They might change the unit partitioning as a side effect.

Testing locally, I do have a different unit partitioning for full builds of cranelift-codegen. I see no reason to consider this to be anything different that codegen partitioning noise.

**Comment by Mark-Simulacrum** [comment]
Ah, ok. Makes sense. Thanks for investigating!

**Pull request merged**